packages/core/src/components/context-menu/contextMenuTarget.tsx

```diff
@@ -17,45 +18,44 @@ export interface IContextMenuTarget extends React.Component<any, any> {
     onContextMenuClose?(): void;
 }
 
-export function ContextMenuTarget<T extends { prototype: IContextMenuTarget }>(constructor: T) {
-    const { render, renderContextMenu, onContextMenuClose } = constructor.prototype;
-
-    if (!isFunction(renderContextMenu)) {
+export function ContextMenuTarget<T extends IConstructor<IContextMenuTarget>>(WrappedComponent: T) {
+    if (!isFunction(WrappedComponent.prototype.renderContextMenu)) {
         console.warn(CONTEXTMENU_WARN_DECORATOR_NO_METHOD);
     }
 
-    // patching classes like this requires preserving function context
-    // tslint:disable-next-line only-arrow-functions
-    constructor.prototype.render = function(this: IContextMenuTarget) {
-        /* tslint:disable:no-invalid-this */
-        const element = render.call(this) as JSX.Element;
+    return class ContextMenuTargetClass extends WrappedComponent {
+        public static displayName = `ContextMenuTarget(${getDisplayName(WrappedComponent)})`;
 
-        if (element == null) {
-            // always return `element` in case caller is distinguishing between `null` and `undefined`
-            return element;
-        }
+        public render() {
+            // TODO handle being applied on a Component that doesn't return an actual Element
+            const element = super.render() as JSX.Element;
 
-        const oldOnContextMenu = element.props.onContextMenu as React.MouseEventHandler<HTMLElement>;
-        const onContextMenu = (e: React.MouseEvent<HTMLElement>) => {
-            // support nested menus (inner menu target would have called preventDefault())
-            if (e.defaultPrevented) {
-                return;
+            if (element == null) {
+                // always return `element` in case caller is distinguishing between `null` and `undefined`
+                return element;
             }
 
-            if (isFunction(this.renderContextMenu)) {
-                const menu = this.renderContextMenu(e);
-                if (menu != null) {
-                    const htmlElement = ReactDOM.findDOMNode(this);
-                    const darkTheme = htmlElement != null && isDarkTheme(htmlElement);
-                    e.preventDefault();
-                    ContextMenu.show(menu, { left: e.clientX, top: e.clientY }, onContextMenuClose, darkTheme);
+            const oldOnContextMenu = element.props.onContextMenu as React.MouseEventHandler<HTMLElement>;
+            const onContextMenu = (e: React.MouseEvent<HTMLElement>) => {
+                // support nested menus (inner menu target would have called preventDefault())
+                if (e.defaultPrevented) {
+                    return;
```

Comment:
should we invoke `oldOnContextMenu` in this code path?

Reply:
Not sure about the logic here; @giladgray ?

Reply:
I'm not sure either. could just change next condition to be `if (isFunction && !prevented)`.